### PR TITLE
Swap tracks example

### PR DESCRIPTION
**Pull request by cretz**
Example showing how to simulate `RTPSender.ReplaceTrack` as mentioned in #1020

### CODE REVIEW

**Comment by Sean-Der** [comment]
This is fantastic! Thank you so much @cretz :)

Merging now

File: examples/swap-tracks/main.go [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+	if err != nil {
+		panic(err)
+	}
+
+	// Output the answer in base64 so we can paste it in browser
+	fmt.Printf("Paste below base64 in browser:\n%v\n", signal.Encode(answer))
+
+	// Asynchronously take all packets in the channel and write them out to our
+	// track
+	go func() {
```
Comment (Sean-Der) [inline examples/swap-tracks/main.go:165]: What do you think of dropping the channel communication and doing the input/output in the OnTrack read loop? The indirection might confuse people new to Go (who don't know channels yet)

Conceptually I think I see two code blocks

* Firing something every 5 seconds 
* Reading the packets, updating the SSRC/SequenceNumber and pushing

File: examples/swap-tracks/main.go [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+				onMe = false
+			}
+		}
+	})
+	// Set the handler for ICE connection state and update chan if connected
+	connected := make(chan struct{}, 1)
+	peerConnection.OnICEConnectionStateChange(func(connectionState webrtc.ICEConnectionState) {
+		fmt.Printf("Connection State has changed %s \n", connectionState.String())
+		// Non-blocking just in case already sent
+		if connectionState == webrtc.ICEConnectionStateConnected {
```
Comment (Sean-Der) [inline examples/swap-tracks/main.go:140]: I would do a context, I think you can call done as many times as you want?

**Comment by Sean-Der** [comment]
This is really amazing @cretz 

I don't very often see demos in WebRTC that makes me go 'Damn that is cool', but this does :) the animation is so fluid, and the switching is so perfect. Really fantastic work.

I am excited to share this on Twitter, you have a lot of cool concepts going on here that people should see. Would you be interested in joining the Pion org? We have a talked a few times, but seeing this really makes me want to get you more involved :)

File: examples/swap-tracks/main.go (hunk 1 above)
Comment (cretz) [reply examples/swap-tracks/main.go:165]: This is what I had originally, but keeping the shared timestamp and sequence numbers intact w/ atomic updates from either side became a bit confusing and not as guaranteed to keep packets (or I'd have to lock the entire loop iteration). But I can change it back to that if I must,.

In this use case, I think the benefit using the channel to combine multiple sources outweighs the cost of somehow a Go dev not knowing about them.

File: examples/swap-tracks/main.go (hunk 2 above)
Comment (cretz) [reply examples/swap-tracks/main.go:140]: There are a ton of ways to skin this cat. Could use a context cancel func, could use an atomic value, could use `sync.Once` with simple chan with `Do(func() { close(doneCh) })`, could use `sync.Once` with `sync.WaitGroup` and `Do(wg.Done)`, could use `sync.Cond`, etc, etc.

If it's much clearer with context cancellation, I can use that (I tend not to use it as a notifier as it does read as clear as context's common uses).

**Comment by cretz** [comment]
> Would you be interested in joining the Pion org? We have a talked a few times, but seeing this really makes me want to get you more involved :)

I am in awe of your dedication, but I fear I don't share it :-) I come back to this project every so often, make a toy, move on, heh. I fear me joining the org would be doing it a disservice as I can just go radio silent here and there. Offer appreciated though.

Here is some code I hacked together for a project I'm working on (untested yet) that lets you switch an inbound stream if it helps anyone:

```go

import (
	"context"
	"io"
	"sync/atomic"

	"github.com/pion/rtp"
	"github.com/pion/webrtc/v2"
)

// TrackSink pipes an incoming track to an outgoing one and allows the incoming
// track to be changed.
type TrackSink struct {
	out            interface{ WriteRTP(*rtp.Packet) error }
	outSSRC        uint32
	inTrack        atomic.Value
	inTrackUpdated chan struct{}
}

// NewTrackSink creates a new TrackSink for the given out track (given as a
// writer and a SSRC value).
func NewTrackSink(out interface{ WriteRTP(*rtp.Packet) error }, outSSRC uint32) *TrackSink {
	return &TrackSink{
		out:     out,
		outSSRC: outSSRC,
		// Buffer one
		inTrackUpdated: make(chan struct{}, 1),
	}
}

type inTrack struct {
	track         *webrtc.Track
	lastTimestamp uint32
}

// SetInTrack changes the current incoming track to the given one. If a nil
// track is given, the outbound writes are stopped until a non-nil track is
// given.
func (t *TrackSink) SetInTrack(track *webrtc.Track) {
	// Change and signal non-blocking
	t.inTrack.Store(&inTrack{track: track})
	select {
	case t.inTrackUpdated <- struct{}{}:
	default:
	}
}

// Run continuously writes to the outbound track as needed until the context is
// closed or an error occurs. If an error occurs on read of a track, onReadError
// is invoked if it's non-nil. Only if the result of onReadError is non-nil will
// the process be stopped and that error returned. Otherwise the read error is
// swalled. All write errors stop this process and are returned. If the context
// is closed, the process is stopped and that error is returned. This always
// returns a non-nil error.
func (t *TrackSink) Run(ctx context.Context, onReadError func(error) error) error {
	// Keep track of the timestamp
	var currTimestamp, lastTimestampDelta uint32
	// The byte buffer and rtp packet to reuse
	buf := make([]byte, 1460)
	var rtpPacket rtp.Packet
	// Keep track of last failed reader to ignore
	var lastFailedIn *inTrack
	// Sequence number to increment each successful packet
	var seqNum uint16
	// Run until context canceled
	for ctx.Err() == nil {
		// Grab in track
		in, _ := t.inTrack.Load().(*inTrack)
		// If the track is not there, wait for signal and try again
		if in == nil || in.track == nil || in == lastFailedIn {
			select {
			case <-ctx.Done():
				return ctx.Err()
			case <-t.inTrackUpdated:
				continue
			}
		}
		// Read info byte buf
		// TODO: What about deadlines?
		n, err := in.track.Read(buf)
		// If no error, unmarshal into packet. We confirmed only the following two
		// header values need to be cleared.
		if err == nil {
			rtpPacket.ExtensionProfile, rtpPacket.ExtensionPayload = 0, nil
			err = rtpPacket.Unmarshal(buf[:n])
		}
		// If there's an error, notify, mark failed, continue
		if err != nil {
			// Send read error
			if onReadError != nil {
				if err = onReadError(err); err != nil {
					return err
				}
			}
			// Mark as last failed and continue
			lastFailedIn = in
			continue
		}
		// Set the ssrc and sequence number
		rtpPacket.SSRC = t.outSSRC
		seqNum++
		rtpPacket.SequenceNumber = seqNum
		// Apply last timestamp delta if there is a prev timestamp. If there's not,
		// it means it's the first time we've seen this track, so we use the last
		// delta from the previous stream.
		if in.lastTimestamp != 0 {
			lastTimestampDelta = rtpPacket.Timestamp - in.lastTimestamp
		}
		// Apply the last timestamp
		currTimestamp += lastTimestampDelta
		in.lastTimestamp = rtpPacket.Timestamp
		rtpPacket.Timestamp = currTimestamp
		// Attempt write. We ignore ErrClosedPipe. We have confirmed that the packet
		// is not referenced beyond the call.
		if err := t.out.WriteRTP(&rtpPacket); err != nil && err != io.ErrClosedPipe {
			return err
		}
	}
	return ctx.Err()
}
```

**Comment by Sean-Der** [comment]
Merged with 97c8f5290c0c00bb4fa7cab989f5a4e9631293fb

No worries about joining, when you get the bug though and want to get deep into jitterbuffer stuff I am here :p 

I ended up dropping the channel completely, I just continue if `trackCount == 0`! I ended up leaving the reader goroutine, I think that is good. I might reorganize things, maybe it would feel better if it right next to the `OnTrack`? I tend to overthink things, but these examples are probably the most important thing we have, lots of people launch their projects off of them :)

**Pull request closed without merging**